HHS/Head-Start-TTADP
# [TTAHUB-2331] TR goal creates goal on RTR

## Description of change

- Add new join table `EventReportPilotGoals` to track goals that have been created from a TR. Stores `goalId`, `eventId`, `sessionId` and `grantId`.
- `SessionReport` hooks:
  - `createGoalsForSessionRecipientsIfNecessary` (`afterCreate` & `afterUpdate`) - does what you probably think it does.
  - `destroyAssociatedGoals` (`beforeDestroy`)
- `EventReport` hooks:
  - `updateGoalText` (`afterUpdate`) - if Goal.name changes, updates Goal.name for all goals created from the TR, unless the session is complete.
- `getGoalsByActivityRecipient` has some logic to exclude returning a Goal when it is associated with a session that is not yet "Complete".

## How to test

This query will be helpful for validating each of these steps:

```sql
SELECT e.id AS "ERPG id",
       erp.data->>'goal' as "EventReportPilot.data->>'goal'",
       g.id AS "Goal.id",
       g.name AS "Goal.name",
       g."grantId" AS "Goal.grantId",
       g.status AS "Goal.status",
       g."previousStatus" AS "Goal.previousStatus",
       s.id AS "SessionReportPilot.id",
       s.data->>'status' AS "SessionReportPilot.data->>'status'"
FROM "EventReportPilotGoals" as e
LEFT JOIN "EventReportPilots" as erp ON e."eventId" = erp.id
LEFT JOIN "Goals" as g ON e."goalId" = g.id
LEFT JOIN "SessionReportPilots" as s ON e."sessionId" = s.id
WHERE e."goalId" IS NOT NULL;
```

- Create an event with text in the "Event goal" field.
- Ensure no goal was created (because no grantId to reference yet).
- Create a session for this event by adding one recipient.
- Run query, ensure a Goal was created and is in Draft and does not appear on the RTR.
- Edit session, add another recipient. Ensure another Goal is created and is also Draft.
- Change the goal text on the event page, run query, ensure "Goal.name" is also updated to match the "event data.goal" text.
- Complete the session, run query, ensure status is now "In Progress" and:
  - Changes to "Event goal" do not change "Goal.name" any longer.
  - The goal appears on the RTR and can be edited.
  - Creating a new session with new recipients will create a Goal with "In Progress" status by default.
- Creating a new session with the same recipient(s) as another session **does not create another Goal for that recipient**.


## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-2331


## Checklists

### Every PR



### Before merge to main


### Production Deploy


### After merge/deploy
